src/value/native_reader.rs

```diff
@@ -115,9 +125,12 @@ impl<R: RawReader> NativeElementIterator<R> {
     fn materialize_struct(&mut self) -> IonResult<OwnedStruct> {
         let mut child_elements = Vec::new();
         self.reader.step_in()?;
-        while let Some(element) = self.materialize_next()? {
+        while let StreamItem::Value(_) | StreamItem::Null(_) = self.reader.next()? {
             let field = self.reader.field_name()?;
-            child_elements.push((owned::text_token(field.as_ref()), element));
+            let value = self
+                .materialize_current()?
+                .expect("materialize_current() returned None for user data");
```

Comment:
> This expect panics because there was no value attached to given field_name?

No, it `expect`s because we've already proven that there was a nullable value and that there weren't any errors while reading:

```rust
        // v-- Here we confirm that there's definitely either a value or a null.
        while let StreamItem::Value(_) | StreamItem::Null(_) = self.reader.next()? {
            let field = self.reader.field_name()?;
            child_elements.push((owned::text_token(field.as_ref()), element));
            let value = self
                .materialize_current()? // <-- here we use `?` to handle any errors
                // At this point, there's definitely a value (or null) and there weren't any errors...
                .expect("materialize_current() returned None for user data");
                // ...so we can use `expect` to dismiss the `None` case.
```

I agree that the error message could be clearer; I'll fix it in a follow-on PR to avoid complicating the rebases I need to do in other PRs.